src/marker.js

```diff
@@ -122,30 +162,12 @@ function applyMarkersColor(path, {markerStart, markerMid, markerEnd, stroke}, st
         node.setAttribute("id", id);
         iriByColor.set(color, (iri = `url(#${id})`));
       }
-      return iri;
+      this.setAttribute(name, iri);
     };
   }
 
-  if (!(markerStart || markerMid || markerEnd)) return;
-
-  const start = markerStart && applyMarker(markerStart);
-  const mid = markerMid && applyMarker(markerMid);
-  const end = markerEnd && applyMarker(markerEnd);
-  if (Z) {
-    for (const g of group(
-      path.filter((i) => i.length >= 2),
-      (d) => Z[select(d).datum()[0]]
-    ).values()) {
-      if (start) select(g.at(0)).attr("marker-start", start);
-      if (mid) {
-        selectAll(g.slice(1)).attr("marker-start", mid);
-        selectAll(g).attr("marker-mid", mid);
-      }
-      if (end) select(g.at(-1)).attr("marker-end", end);
-    }
-  } else {
-    if (start) path.attr("marker-start", start);
-    if (mid) path.attr("marker-mid", mid);
-    if (end) path.attr("marker-end", end);
-  }
+  if (markerStart) path.each(applyMarker("marker-start", markerStart, orient && ((i) => orient(i) & START)));
+  if (markerMid && orient) path.each(applyMarker("marker-start", markerMid, (i) => orient(i) ^ START));
```

Comment:
```suggestion
  if (markerMid && orient) path.each(applyMarker("marker-start", markerMid, (i) => !(orient(i) & START)));
```

(to not evaluate to END when a segment is START+END = 3)